File: SiderShopping/app/src/main/java/com/mmart/sidershopping/MainActivity.java
package com.mmart.sidershopping;

import android.support.design.widget.TabLayout;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;

import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ListView;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {
    private String[] imageUrls = new String[]{
            "https://cdn.pixabay.com/photo/2016/11/11/23/34/cat-1817970_960_720.jpg",
            "https://cdn.pixabay.com/photo/2017/12/21/12/26/glowworm-3031704_960_720.jpg",
            "https://cdn.pixabay.com/photo/2017/12/24/09/09/road-3036620_960_720.jpg",
            "https://cdn.pixabay.com/photo/2017/11/07/00/07/fantasy-2925250_960_720.jpg",
            "https://cdn.pixabay.com/photo/2017/10/10/15/28/butterfly-2837589_960_720.jpg"
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ViewPager viewPager = findViewById(R.id.view_pager);
        ViewPagerAdapter adapter = new ViewPagerAdapter(this, imageUrls);

        TabLayout tabLayout = (TabLayout) findViewById(R.id.tabDots);
        tabLayout.setupWithViewPager(viewPager, true);

        viewPager.setAdapter(adapter);

        setCategories();
        setBestSellers();
    }

    private void setBestSellers() {
        RecyclerView bestSellerList = (RecyclerView) findViewById(R.id.best_seller_list);
        LinearLayoutManager layoutManagerBestSeller = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
        bestSellerList.setLayoutManager(layoutManagerBestSeller);

        ArrayList<BestSeller> bestSellers = new ArrayList<BestSeller>();

        bestSellers.add(new BestSeller(1,
                "Produto 1",
                "De: 9999,99",
                "Por: 9998,99",
                "https://images-americanas.b2w.io/produtos/01/00/item/133204/8/133204824_1GG.jpg"));

        bestSellers.add(new BestSeller(1,
                "Produto 1",
                "De: 9999,99",
                "Por: 9998,99",
                "https://images-americanas.b2w.io/produtos/01/00/item/133204/8/133204824_1GG.jpg"));

        bestSellers.add(new BestSeller(1,
                "Produto 1",
                "De: 9999,99",
                "Por: 9998,99",
                "https://images-americanas.b2w.io/produtos/01/00/item/133204/8/133204824_1GG.jpg"));

        bestSellers.add(new BestSeller(1,
                "Produto 1",
                "De: 9999,99",
                "Por: 9998,99",
                "https://images-americanas.b2w.io/produtos/01/00/item/133204/8/133204824_1GG.jpg"));

        bestSellers.add(new BestSeller(1,
                "Produto 1",
                "De: 9999,99",
                "Por: 9998,99",
                "https://images-americanas.b2w.io/produtos/01/00/item/133204/8/133204824_1GG.jpg"));

        BestSellerAdapter bestSellerAdapter = new BestSellerAdapter(this, bestSellers);
        bestSellerList.setAdapter(bestSellerAdapter);
    }

    private void setCategories() {
        ArrayList<Category> categories = new ArrayList<Category>();

        RecyclerView lista = (RecyclerView) findViewById(R.id.lista);
        LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
        lista.setLayoutManager(layoutManager);

        categories.add(new Category(1, "Categoria 1", "https://images-americanas.b2w.io/produtos/01/00/item/133204/8/133204824_1GG.jpg"));
        categories.add(new Category(2, "Categoria 2", "https://images-americanas.b2w.io/produtos/01/00/item/133204/8/133204824_1GG.jpg"));
        categories.add(new Category(3, "Categoria 3", "https://images-americanas.b2w.io/produtos/01/00/item/133204/8/133204824_1GG.jpg"));
        categories.add(new Category(4, "Categoria 4", "https://images-americanas.b2w.io/produtos/01/00/item/133204/8/133204824_1GG.jpg"));
        categories.add(new Category(5, "Categoria 5", "https://images-americanas.b2w.io/produtos/01/00/item/133204/8/133204824_1GG.jpg"));
        categories.add(new Category(6, "Categoria 6", "https://images-americanas.b2w.io/produtos/01/00/item/133204/8/133204824_1GG.jpg"));

        CategoryAdapter categoryAdapter = new CategoryAdapter(categories, this);
        lista.setAdapter(categoryAdapter);
    }
}

File: app/src/main/java/br/com/amedigital/MainActivity.kt
package br.com.amedigital

import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import br.com.amedigital.banner.*
import br.com.amedigital.category.CategoryAdapter
import br.com.amedigital.category.CategoryContract
import br.com.amedigital.category.CategoryPresenter
import br.com.amedigital.model.Banner
import br.com.amedigital.model.Category
import br.com.amedigital.model.Product
import br.com.amedigital.network.LojinhaServiceEndPoint
import br.com.amedigital.product.*
import br.com.amedigital.product.ProductByCategoryActivity.Companion.CATEGORY
import br.com.amedigital.util.NavigationViewController
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity(), BannerContract.View,
                                            CategoryContract.View,
                                            ProductContract.View{

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        loadBanners()
        loadCategories()
        loadProductsMoreSales()
        navigationDrawer()
    }

    private fun navigationDrawer() {
        val navigationViewController = NavigationViewController(this, toolbar)
        val menu = navigationView.menu.findItem(R.id.drawer_home)
        menu.isChecked = true
        navigationViewController.initNavigationDrawer(navigationView, drawer, this@MainActivity)
        navigationView.inflateHeaderView(R.layout.drawer_header)
    }

    private fun loadBanners() {
        val call = LojinhaServiceEndPoint().getBanner()
        val presenter = BannerPresenter(call, this)
        presenter.loadBanners()
    }

    private fun loadCategories() {
        val call = LojinhaServiceEndPoint().getCategory()
        val presenter = CategoryPresenter(call, this)
        presenter.loadCategories()
    }

    private fun loadProductsMoreSales() {
        val call = LojinhaServiceEndPoint().getTopSellingProducts()
        val presenter = ProductPresenter()
        presenter.loadProductsMoreSales(call,this)
    }

    override fun setProgressIndicatorBanner(active: Boolean) {
        progressBarBanner.visibility = if (active) View.VISIBLE else View.GONE
    }

    override fun showBanners(banners: ArrayList<Banner>) {
        bannerSlider.setAdapter(BannerAdapter(banners))
        bannerSlider.setInterval(5000)
    }

    override fun showErrorBanner(message: String) {
        showErrorMessages(message)
    }

    override fun setProgressIndicatorCategory(active: Boolean) {
        progressBarCategory.visibility = if (active) View.VISIBLE else View.GONE
    }

    override fun showCategories(categories: ArrayList<Category>) {
        recyclerviewCategory?.adapter = CategoryAdapter(categories) {
            val category = categories[it]
            val intent = Intent(this@MainActivity, ProductByCategoryActivity::class.java)
            intent.putExtra(CATEGORY, category)
            startActivity(intent)
        }
        val layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        recyclerviewCategory?.layoutManager = layoutManager
    }

    override fun showErrorCategory(message: String) {
        showErrorMessages(message)
    }

    override fun setProgressProductMore(active: Boolean) {
        progressBarProdMore.visibility = if (active) View.VISIBLE else View.GONE
    }

    override fun showProductsMore(products: ArrayList<Product>) {
        recyclerviewMoreSales?.adapter = ProductAdapter(products) {
            val intent = Intent(this@MainActivity, ProductDetailActivity::class.java)
            intent.putExtra(ProductDetailActivity.PRODUCT, products[it])
            startActivity(intent)
        }
        val layoutManager = LinearLayoutManager(this)
        recyclerviewMoreSales?.layoutManager = layoutManager
    }

    override fun showEmpytProductMore(message: String) {
        textViewEmpytMore.visibility = View.VISIBLE
        textViewEmpytMore.text = message
    }

    override fun showErrorProductMore(message: String) {
        showErrorMessages(message)
    }

    private fun showErrorMessages(errorMessage : String) {
        val snackBarError = Snackbar.make(textViewCategoria, errorMessage, Snackbar.LENGTH_INDEFINITE)
        snackBarError.setAction(getString(R.string.ok)) { snackBarError.dismiss() }
        snackBarError.setActionTextColor(Color.WHITE)
        snackBarError.show()
    }
}
